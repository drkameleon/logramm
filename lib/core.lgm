/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/core.lgm
 **********************************************************/

//================================================
// Module definition
//================================================

Core :: 
{

	//================================================
	// External D-bindings
	//================================================

	// String ::

	external contains(str,what) # [
		.params		: "string,string",
		.returns    : "boolean",
		.help		: "Checks if string contains 'what'"
	] => "string.contains";

	external length(str) # [
		.params		: "string",
		.returns 	: "number",
		.help		: "Returns length of string"
	] => "string.length";

	external replace(str,what,repl) # [
		.params		: "string,string,string",
		.returns    : "string",
		.help		: "Replace all occurences of 'what' with 'repl' and returns string"
	] => "string.replace";

	external split(str,delim) # [
		.params		: "string,string",
		.returns    : "array",
		.help		: "Split string by 'delim' and return array"
	] => "string.split";

	// Array ::

	external count(arr) # [
		.params		: "array", 
		.returns    : "number", 
		.help		: "Returns number of elements in array"
	] => "array.count";

	// Dictionary ::

	external keys(dict) # [
		.params		: "dictionary", 
		.returns    : "array", 
		.help		: "Returns dictionary keys as an array"
	] => "dictionary.keys";

	//================================================
	// Functions
	//================================================

	// Array ::

	contains(arr,item) # [
		.params		: "array,any", 
		.returns    : "boolean", 
		.help		: "Checks if array contains 'item'"
	] =>
	{
		@ i->arr : i==item :==> true;

		==> false;
	}

	join(arr,delim) # [
		.params		: "array,string",
		.returns    : "string",
		.help		: "Returns string by joining string in 'arr' using 'delim'"
	] =>
	{
		ret = "";

		i = 0;

		@ i<count(arr) : 
		{
			ret += arr[i];

			i!=count(arr)-1 : ret += delim;

			i += 1;
		}

		==> ret;
	}

	// Dictionary ::

	values(dict) # [
		.params		: "dictionary",
		.returns    : "array",
		.help		: "Returns an array containing dictionary's values"
	] =>
	{
		keys = keys(dict);

		keyIndex = 0;
		ret = [];

		@ keyIndex<count(keys) : 
		{
			ret += dict[keys[keyIndex]];
			keyIndex += 1;
		}

		==> ret;
	}
}

