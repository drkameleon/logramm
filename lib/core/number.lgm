/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/core/number.lgm
 **********************************************************/

//================================================
// Module definition
//================================================

Number :: 
{

	//================================================
	// External D-bindings
	//================================================

	external fromBase(str,n) # [
		.params		: "string,number",
		.returns 	: "number",
		.help 		: "Returns number, given its base-n representation in string"
	] => "math.fromBase";

	external sin(x) # [
		.params		: "number",
		.returns 	: "number",
		.help 		: "Returns sine of 'x'"
	] =>  "math.sin";

	external cos(x) # [
		.params		: "number",
		.returns 	: "number",
		.help 		: "Returns cosine of 'x'"
	] =>  "math.cos";

	external tan(x) # [
		.params		: "number",
		.returns 	: "number",
		.help 		: "Returns tangent of 'x'"
	] =>  "math.tan";

	external sinh(x) # [
		.params		: "number",
		.returns 	: "number",
		.help 		: "Returns hyperbolic sine of 'x'"
	] =>  "math.sinh";

	external cosh(x) # [
		.params		: "number",
		.returns 	: "number",
		.help 		: "Returns hyperbolic cosine of 'x'"
	] =>  "math.cosh";

	external tanh(x) # [
		.params		: "number",
		.returns 	: "number",
		.help 		: "Returns hyperbolic tangent of 'x'"
	] =>  "math.tanh";

	external arcsin(x) # [
		.params		: "number",
		.returns 	: "number",
		.help 		: "Returns arc sine of 'x'"
	] =>  "math.arcsin";

	external arccos(x) # [
		.params		: "number",
		.returns 	: "number",
		.help 		: "Returns arc cosine of 'x'"
	] =>  "math.arccos";

	external arctan(x) # [
		.params		: "number",
		.returns 	: "number",
		.help 		: "Returns arc tangent of 'x'"
	] =>  "math.arctan";

	external arcsinh(x) # [
		.params		: "number",
		.returns 	: "number",
		.help 		: "Returns inverse hyperbolic sine of 'x'"
	] =>  "math.arcsinh";

	external arccosh(x) # [
		.params		: "number",
		.returns 	: "number",
		.help 		: "Returns inverse hyperbolic cosine of 'x'"
	] =>  "math.arccosh";

	external arctanh(x) # [
		.params		: "number",
		.returns 	: "number",
		.help 		: "Returns inverse hyperbolic tangent of 'x'"
	] =>  "math.arctanh";

	external floor(x) # [
		.params		: "number",
		.returns 	: "number",
		.help 		: "Returns 'x', rounded downwards"
	] =>  "math.floor";

	external ceil(x) # [
		.params		: "number",
		.returns 	: "number",
		.help 		: "Returns 'x', rounded upwards"
	] =>  "math.ceil";

	external round(x) # [
		.params		: "number",
		.returns 	: "number",
		.help 		: "Returns 'x', rounded to the nearest integer"
	] =>  "math.round";

	external ln(x) # [
		.params		: "number",
		.returns 	: "number",
		.help 		: "Returns the natural logarithm of 'x'"
	] =>  "math.ln";

	external log(x) # [
		.params		: "number",
		.returns 	: "number",
		.help 		: "Returns the base-10 logarithm of 'x'"
	] =>  "math.log";

	external pow(x,y) # [
		.params		: "number,number",
		.returns 	: "number",
		.help 		: "Returns 'x' raised to the power of 'y'"
	] =>  "math.pow";

	external exp(x) # [
		.params		: "number",
		.returns 	: "number",
		.help 		: "Returns exponential of 'x'"
	] =>  "math.exp";

	external sqrt(x) # [
		.params		: "number",
		.returns 	: "number",
		.help 		: "Returns square root of 'x'"
	] =>  "math.sqrt";

	external trunc(x) # [
		.params		: "number",
		.returns 	: "number",
		.help 		: "Returns integer portion of 'x'"
	] =>  "math.trunc";

	external E() # [
		.params		: "no",
		.returns    : "number",
		.help		: "Returns E constant"
	] => "math.e";

	external Pi() # [
		.params		: "no",
		.returns 	: "number",
		.help 		: "Returns Pi constant"
	] => "math.pi";

	//================================================
	// Functions
	//================================================

	fromBin(str) # [
		.params		: "string",
		.returns 	: "number",
		.help 		: "Returns number, given its binary representation in string"
	] => fromBase(str,2);

	fromHex(str) # [
		.params		: "string",
		.returns 	: "number",
		.help 		: "Returns number, given its hexadecimal representation in string"
	] => fromBase(str,16);

	fromOct(str) # [
		.params		: "string",
		.returns 	: "number",
		.help 		: "Returns number, given its octal representation in string"
	] => fromBase(str,8);

	fact(x) # [
		.params		: "number",
		.returns    : "number",
		.help		: "Returns the factorial of 'x'"
	] =>
	{
		x==1    :==> 1;
		else    :==> x * fact(x-1);
	}

	fibo(x) # [
		.params		: "number",
		.returns    : "number",
		.help		: "Returns nth fibonacci number"
	] => 
	{
		x==0 :==> 1;
		else : 
			x==1 :==> 1;
			else :==> fibo(x-1) + fibo(x-2);
	}

	abs(x) # [
		.params		: "number",
		.returns    : "number",
		.help		: "Returns the absolute value of 'n'"
	] =>
	{
		x>=0    :==> x;
		else	:==> -x;
	}

}
