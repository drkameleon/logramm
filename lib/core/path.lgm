/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/core/path.lgm
 **********************************************************/

//================================================
// Module definition
//================================================

Path :: 
{

	//================================================
	// External D-bindings
	//================================================

	external isFile(path) # [
		.params		: "string", 
		.returns    : "boolean", 
		.help		: "Checks if file at 'path' is a file"
	] => "path.isFile";

	external isDirectory(path) # [
		.params		: "string", 
		.returns    : "boolean", 
		.help		: "Checks if file at 'path' is a directory"
	] => "path.isDirectory";

	external isSymlink(path) # [
		.params		: "string", 
		.returns    : "boolean", 
		.help		: "Checks if file at 'path' is a symlink"
	] => "path.isSymlink";

	external createDirectory(path) # [
		.params		: "string", 
		.returns    : "no", 
		.help		: "Creates directory at 'path'"
	] => "path.createDirectory";

	external currentDirectory()	# [
		.params		: "", 
		.returns    : "string", 
		.help		: "Returns current working directory"
	] => "path.currentDirectory";

	external directoryContents(dir)	# [
		.params		: "string", 
		.returns    : "array", 
		.help		: "Returns array with contents of given directory"
	] => "path.directoryContents";

	external directorySeparator() # [
		.params		: "",
		.returns 	: "string",
		.help 		: "Returns symbol used to separate directory names in a path"
	] => "path.directorySeparator";

	external filename(path) # [
		.params		: "string",
		.returns 	: "string",
		.help 		: "Returns name of file, in file path, without leading directories"
	] => "path.filename";

	external directory(path) # [
		.params 	: "string",
		.returns 	: "string",
		.help		: "Returns the directory part of given path"
	] => "path.directory";

	external extension(path) # [
		.params		: "string",
		.returns 	: "string",
		.help 		: "Returns extension part of given path, including the dot"
	] => "path.extension";

	external withoutExtension(path) # [
		.params		: "string",
		.returns 	: "string",
		.help 		: "Returns filepath by removing extension"
	] => "path.withoutExtension";

	external normalized(path) # [
		.params 	: "string",
		.returns 	: "string",
		.help 		: "Returns path, after resolving current/parent directory symbols"
	] => "path.normalized";

	external expandTilde(path) # [
		.params 	: "string",
		.returns 	: "string",
		.help 		: "Returns path, after expanding tilde"
	] => "path.expandTilde";

}
