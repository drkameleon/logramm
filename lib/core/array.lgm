/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/core/array.lgm
 **********************************************************/

//================================================
// Imported modules
//================================================

import core;

//================================================
// Module definition
//================================================

Array ::
{

	//================================================
	// Functions
	//================================================

	itemAtIndex(arr,index) # [
		.params		: "array,number",
		.returns    : "any",
		.help		: "Returns item of array at index"
	] => arr[index];
	

	prependItem(arr,item) # [
		.params		: "array,any",
		.returns    : "array",
		.help		: "Returns resulting array after prepending 'item'"
	] => [item] + arr;

	appendItem(arr,item) # [
		.params		: "array,any",
		.returns    : "array",
		.help		: "Returns resulting array after appending 'item'"
	] => arr + item;

	insertItemAtIndex(arr,index,item) # [
		.params		: "array,number,any",
		.returns    : "array",
		.help		: "Returns resulting array after inserting 'item' at 'index'"
	] => arr[..index] + item + arr[index..];

	setItemAtIndex(arr,index,item) # [
		.params		: "array,number,any",
		.returns    : "array",
		.help		: "Returns resulting array after setting item at 'index' to 'item'"
	] => 
	{
		index+1<=count(arr)-1 :==> arr[..index] + item + arr[index+1..];
		index+1>count(arr)-1  :==> arr[..index] + item;
	}

	removeItemAtIndex(arr,index) # [
		.params		: "array,number",
		.returns    : "array",
		.help		: "Returns resulting array after removing item at 'index'"
	] =>
	{
		index+1<=count(arr)-1 :==> arr[..index] + arr[index+1..];
		index+1>count(arr)-1  :==> arr[..index];
	}

	removeItem(arr,item) # [
		.params		: "array,any",
		.returns    : "array",
		.help		: "Returns resulting array after removing 'item'"
	] =>
	{
		newArr = [];

		@ i->arr : i!=item : newArr += i;

		==> newArr;
	}

	first(arr) # [
		.params		: "array",
		.returns    : "any",
		.help		: "Returns first item in array"
	] => arr[0];
	
	last(arr) # [
		.params		: "array",
		.returns    : "any",
		.help		: "Returns last item in array"
	] => arr[count(arr)-1];
	
	slice(arr, begin, end) # [
		.params		: "array,number,number",
		.returns    : "array",
		.help		: "Returns a slice of the array from 'begin' up to 'end'"
	] => arr[begin..end];

	push(arr,item) # [
		.params		: "array,any",
		.returns    : "array",
		.help		: "Returns resulting array after pushing 'item'"
	] => arr + item;

	pop(arr) # [
		.params		: "array",
		.returns    : "any",
		.help		: "Pops and returns last item from array"
	] => 
	{
		lastItem = arr[count(arr)-1];
		arr = arr[..count(arr)-1];
		==> lastItem;
	}
	
	merge(arr1,arr2) # [
		.params		: "array,array",
		.returns    : "array",
		.help		: "Returns resulting array after merging 'arr1' and 'arr2'"
	] => arr1 + arr2;
	
	reverse(arr) # [
		.params		: "array",
		.returns    : "array",
		.help		: "Returns resulting array after reversing its elements"
	] => 
	{
		newArr = [];

		i = count(arr)-1;
		@ i>=0 : {
			newArr += arr[i];
			i -= 1;
		}

		==> newArr;
	}
	
	isEmpty(arr) # [
		.params		: "array", 
		.returns    : "boolean", 
		.help		: "Checks if array is empty"
	] => 
	{
		count(arr)==0 :==> true;
		else 		  :==> false;
	}

	contains(arr,item) # [
		.params		: "array,any", 
		.returns    : "boolean", 
		.help		: "Checks if array contains 'item'"
	] =>
	{
		@ i->arr : i==item :==> true;

		==> false;
	}

	removeDuplicates(arr) # [
		.params		: "array",
		.returns    : "array",
		.help		: "Returns resulting array after removing duplicate elements"
	] =>
	{
		newArr = [];

		@ i->arr : contains(newArr,i)==false : newArr += i;

		==> newArr;
	}

	map(arr,pattern) # [
		.params		: "array,string",
		.returns    : "array",
		.help		: "Returns resulting array after items have been mapped using 'pattern'"
	] =>
	{
		newArr = [];
		func = "newItem = " + replace(pattern,"$","item") + ";";

		@ item->arr :
		{
			exec func;
			newArr += newItem;
		}

		==> newArr;
	}

	filter(arr,pattern) # [
		.params		: "array,string",
		.returns    : "array",
		.help		: "Returns resulting array after items have been filtered using 'pattern'"
	] =>
	{
		newArr = [];
		func = replace(pattern,"$","item") + " : newArr += item;";

		@ item->arr : exec func;

		==> newArr;
	}

	sort(arr) # [
		.params		: "array",
		.returns 	: "array",
		.help 		: "Returns array after sorting its elements"
	] => 
	{
		n = count(arr);
		swapped = true;

		@ (swapped) :
		{
			swapped = false;

			@ i -> [1..n] :
			{
				arr[i-1] > arr[i] :
				{
					temp = arr[i-1]; 
					arr[i-1] = arr[i];
					arr[i] = temp;

					swapped = true;
				}
			}
		}

		==> arr;
	} 

	search(arr,item) # [
		.params		: "array,any",
		.returns 	: "number",
		.help 		: "Searches for item in array and returns its index, or -1 if not found"
	] =>
	{
		@ index,i -> arr :
			i == item :==> index;

		==> -1;
	}

	allOf(arr,value) # [
		.params		: "array,any",
		.returns 	: "boolean",
		.help		: "Checks if all of array's items are equal to 'value'"
	] =>
	{
		@ item-> arr :
			item!=value :==> false;

		==> true;
	}

	min(arr) # 	[
		.params		: "array",
		.returns    : "number",
		.help		: "Returns minimum element in array"
	] =>
	{
		min = 2000000000;

		@ item->arr :
			item < min : min = item;

		==> min;
	}

	max(arr) # [
		.params		: "array",
		.returns    : "number",
		.help		: "Returns maximum element in array"
	] =>
	{
		max = -2000000000;

		@ item->arr :
			item > max : max = item;

		==> max;
	}
	
	sum(arr) # [
		.params		: "array",
		.returns    : "number",
		.help		: "Returns sum of elements in array"
	] =>
	{
		ret = 0;

		@ x->arr : ret+=x;

		==> ret;
	}

}
