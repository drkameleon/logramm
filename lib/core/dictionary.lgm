/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/core/dictionary.lgm
 **********************************************************/

//================================================
// Imported modules
//================================================

import core;

//================================================
// Module definition
//================================================

Dictionary :: 
{

	//================================================
	// Functions
	//================================================

	itemWithKey(dict,key) # [
		.params		: "dictionary,any",
		.returns    : "any",
		.help		: "Returns item for key 'key'"
	] => dict[key];

	isEmpty(dict) # [
		.params		: "dictionary", 
		.returns    : "boolean", 
		.help		: "Checks if dictionary is empty"
	] =>
	{
		c = count(keys(dict));

		c==0 :==> true;
		c!=0 :==> false;
	}

	setItemWithKey(dict,key,item) # [
		.params		: "dictionary,any,any",
		.returns    : "dictionary",
		.help		: "Returns resulting array after setting item at 'index' to 'item'"
	] => 
	{
		newDict = [:];

		keys = keys(dict);
		ind = 0;

		@ ind<count(keys) : 
		{
			keys[ind]==key : newDict += [ keys[ind] : item ];
			keys[ind]!=key : newDict += [ keys[ind] : dict[keys[ind]] ];

			ind += 1;
		}

		==> newDict;
	}

	deleteItemWithKey(dict,key) # [
		.params		: "dictionary,any",
		.returns    : "dictionary",
		.help		: "Returns resulting dictionary after removing item with key 'key'"
	] =>
	{
		newDict = [:];

		keys = keys(dict);
		ind = 0;

		@ ind<count(keys) : 
		{
			keys[ind]!=key : newDict += [ keys[ind] : dict[keys[ind]] ];

			ind += 1;
		}

		==> newDict;
	}

	size(dict) # [
		.params		: "dictionary", 
		.returns    : "number", 
		.help		: "Returns number of items in dictionary"
	] => count(keys(dict));

	keyExists(dict,key) # [
		.params		: "dictionary,string",
		.returns    : "boolean",
		.help		: "Checks if key exists in dictionary"
	] => contains(keys(dict),key);

}

