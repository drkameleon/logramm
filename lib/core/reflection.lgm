/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/core/reflection.lgm
 **********************************************************/

//================================================
// Imported modules
//================================================

import core;

//================================================
// Module definition
//================================================

Reflection :: 
{

	//================================================
	// External D-bindings
	//================================================

	external symbolExists(symbol) # [
		.params		: "string",
		.returns 	: "boolean",
		.help		: "Checks if the given symbol/variable is defined"
	] => "reflection.symbolExists";

	external methodInfo(func) # [
		.params		: "string",
		.returns	: "dictionary",
		.help		: "Returns info for the given method or false if not found"
	] => "reflection.functionInfo";

	external methodExists(func) # [
		.params		: "string",
		.returns 	: "boolean",
		.help		: "Checks if the given method is defined"
	] => "reflection.functionExists";

	external methodExists(mod,func) # [
		.params		: "string,string",
		.returns 	: "boolean",
		.help		: "Checks if the given method, in module 'mod', is defined"
	] => "reflection.functionExistsInModule";

	external inspect(symbol) # [
		.params		: "any",
		.returns	: "no",
		.help		: "Prints debugging information related to given symbol"
	] => "reflection.inspect";

	external symbols() # [
		.params		: "",
		.returns 	: "array",
		.help		: "Returns an array with defined symbols"
	] => "reflection.symbols";

	external methods() # [
		.params		: "",
		.returns 	: "array",
		.help		: "Returns an array with defined methods"
	] => "reflection.methods";

	//================================================
	// Functions
	//================================================

	callMethod(func) # [
		.params		: "string",
		.returns 	: "boolean",
		.help		: "Executes method by name 'func' and returns true if successful"
	] =>
	{
		methodExists(func) :
		methodInfo(func).params.count() == 0 : { exec "{:func}();"; ==> true; }
		
		==> false;
	}

	callMethod(mod,func) # [
		.params		: "string,string",
		.returns 	: "boolean",
		.help		: "Executes method by name 'func', in module 'mod', and returns true if successful"
	] =>
	{
		methodExists(mod,func) :
		methodInfo(func).params.count() == 0 : { exec "{:mod}::{:func}();"; ==> true; }
		
		==> false;
	}

	callMethodWithArgs(func,args) # [
		.params		: "string,array",
		.returns 	: "boolean",
		.help		: "Executes method by name 'func', using arguments, and returns true if successful"
	] =>
	{
		argstr = args.join(",");

		methodExists(func) :
		methodInfo(func).params.count() == args.count() : { exec "{:func}({:argstr});"; ==> true; }
		
		==> false;
	}

	callMethodWithArgs(mod,func,args) # [
		.params		: "string,string,array",
		.returns 	: "boolean",
		.help		: "Executes method by name 'func', in module 'mod', using arguments, and returns true if successful"
	] =>
	{
		argstr = args.join(",");

		methodExists(mod,func) :
		methodInfo(func).params.count() == args.count() : { exec "{:mod}::{:func}({:argstr});"; ==> true; }
		
		==> false;
	}

	importPackage(pkg) # [
		.params		: "string",
		.returns 	: "no",
		.help		: "Programmatically import package by string name 'pkg'"
	] =>
	{
		exec "import {:pkg};";
	}

}
