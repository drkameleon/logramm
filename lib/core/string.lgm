/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/core/string.lgm
 **********************************************************/

//================================================
// Imported modules
//================================================

import core;

//================================================
// Module definition
//================================================


String :: 
{

	//================================================
	// External D-bindings
	//================================================

	external ucase(str) # [
		.params		: "string",
		.returns    : "string",
		.help		: "Returns a string with all of its characters converted to uppercase"
	] => "string.uppercase";

	external lcase(str) # [
		.params		: "string",
		.returns    : "string",
		.help		: "Returns a string with all of its characters converted to lowercase"
	] => "string.lowercase";

	external trim(str) # [
		.params		: "string",
		.returns    : "string",
		.help		: "Returns a string, having stripped both leading and trailing whitespace"
	] => "string.trim";

	external ltrim(str) # [
		.params		: "string",
		.returns    : "string",
		.help		: "Returns a string, having stripped leading whitespace"
	] => "string.trimLeft";

	external rtrim(str) # [
		.params		: "string",
		.returns    : "string",
		.help		: "Returns a string, having stripped trailing whitespace"
	] => "string.trimRight";

	external find(str,what) # [
		.params		: "string,string",
		.returns    : "number",
		.help		: "Returns index of first occurence of 'what' in string"
	] => "string.find";

	external chop(str) # [
		.params		: "string",
		.returns    : "string",
		.help		: "Returns a string, having removed the last character"
	] => "string.chop";

	external isNumeric(str) # [
		.params		: "string",
		.returns    : "boolean",
		.help		: "Checks if string can be be formatted a number"
	] => "string.isNumeric";

	external soundex(str) # [
		.params		: "string",
		.returns    : "string",
		.help		: "Converts a word into 4 characters based on how the word sounds phonetically, using the Soundex algorithm"
	] => "string.soundex";

	external startsWith(str,what) # [
		.params		: "string,string",
		.returns    : "boolean",
		.help		: "Checks if string begins with 'what'"
	] =>  "string.startsWith";

	external endsWith(str,what) # [
		.params		: "string,string",
		.returns    : "boolean",
		.help		: "Checks if string ends with 'what'"
	] => "string.endsWith";

	external chomp(str) # [
		.params		: "string",
		.returns    : "string",
		.help		: "Returns string after removing trailing newline characters"
	] => "string.chomp";

	external justify(str,padding) # [
		.params		: "string,number",
		.returns 	: "string",
		.help 		: "Returns string after left justifying using padding"
	] => "string.justifyLeft";

	external justifyRight(str,padding) # [
		.params		: "string,number",
		.returns 	: "string",
		.help 		: "Returns string after right justifying using padding"
	] => "string.justifyRight";

	external justifyCenter(str,padding) # [
		.params		: "string,number",
		.returns 	: "string",
		.help 		: "Returns string after center justifying using padding"
	] => "string.justifyCenter";

	//================================================
	// Functions
	//================================================

	isEmpty(str) # [
		.params		: "string",
		.returns    : "boolean",
		.help		: "Checks if string is empty"
	] =>
	{
		c = length(str);

		c==0 :==> true;
		c!=0 :==> false;
	}

	ucaseFirst(str) # [
		.params		: "string",
		.returns    : "string",
		.help		: "Returns string with its first character in uppercase"
	] =>  ucase(str[0]) + str[1..];

	lcaseFirst(str) # [
		.params		: "string",
		.returns    : "string",
		.help		: "Returns string with its first character in lowercase"
	] => lcase(str[0]) + str[1..];

}

