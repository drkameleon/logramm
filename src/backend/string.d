/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** src/backend/string.d
 **********************************************************/

module backend.string;

//================================================
// Imports
//================================================

import std.algorithm;
import std.array;
import std.conv;
import std.regex;
import std.stdio;
import std.string;

import value;

//================================================
// Class
//================================================

class LGM_String
{
	static Value length(Value[] v)
	{
		string str = v[0].content.s;

		return new Value(str.length);
	}

	static Value uppercase(Value[] v)
	{
		string str = v[0].content.s;

		return new Value(str.toUpper);
	}

	static Value lowercase(Value[] v)
	{
		string str = v[0].content.s;

		return new Value(str.toLower);
	}

	static Value trim(Value[] v)
	{
		string str = v[0].content.s;

		return new Value(str.strip());
	}

	static Value trimLeft(Value[] v)
	{
		string str = v[0].content.s;

		return new Value(str.stripLeft());
	}

	static Value trimRight(Value[] v)
	{
		string str = v[0].content.s;

		return new Value(str.stripRight());
	}

	static Value replace(Value[] v)
	{
		string str = v[0].content.s;
		string what = v[1].content.s;
		string repl = v[2].content.s;

		return new Value(str.replace(what,repl));
	}

	static Value replaceFirst(Value[] v)
	{
		string str = v[0].content.s;
		string what = v[1].content.s;
		string repl = v[2].content.s;

		return new Value(str.replaceFirst(what,repl));
	}

	static Value find(Value[] v)
	{
		string str = v[0].content.s;
		string what = v[1].content.s;

		return new Value(str.indexOf(what));
	}

	static Value chop(Value[] v)
	{
		string str = v[0].content.s;

		return new Value(str.chop());
	}

	static Value contains(Value[] v)
	{
		string str = v[0].content.s;
		string what = v[1].content.s;

		return new Value(str.indexOf(what)!=-1);
	}

	static Value isNumeric(Value[] v)
	{
		string str = v[0].content.s;

		return new Value(str.isNumeric());
	}

	static Value split(Value[] v)
	{
		string str = v[0].content.s;
		string delim = v[1].content.s;

		string[] parts = str.splitter(delim).array();
		Value[] ret;

		foreach (string part; parts)
			ret ~= new Value(part);

		return new Value(ret);
	}

	static Value soundex(Value[] v)
	{
		string str = v[0].content.s;

		string ret = "";
		foreach (char c; str.soundex())
			ret ~= c;

		return new Value(ret);
	}

	static Value startsWith(Value[] v)
	{
		string str = v[0].content.s;
		string what = v[1].content.s;

		bool ret;

		if (std.algorithm.startsWith(str,what)==1) ret = true;
		else ret = false;

		return new Value(ret);
	}

	static Value endsWith(Value[] v)
	{
		string str = v[0].content.s;
		string what = v[1].content.s;

		bool ret;

		if (std.algorithm.endsWith(str,what)==1) ret = true;
		else ret = false;

		return new Value(ret);
	}

	static Value chomp(Value[] v)
	{
		string str = v[0].content.s;

		string ret = std.string.chomp(str);

		return new Value(ret);
	}

	static Value justifyLeft(Value[] v)
	{
		string str = v[0].content.s;
		long padding = v[1].content.i;

		string ret = leftJustify(str,to!int(padding));

		return new Value(ret);
	}

	static Value justifyRight(Value[] v)
	{
		string str = v[0].content.s;
		long padding = v[1].content.i;

		string ret = rightJustify(str,to!int(padding));

		return new Value(ret);
	}

	static Value justifyCenter(Value[] v)
	{
		string str = v[0].content.s;
		long padding = v[1].content.i;

		string ret = center(str,to!int(padding));

		return new Value(ret);
	}

	// Regex

	static Value matches(Value[] v)
	{
		string str = v[0].content.s;
		string pattern = v[1].content.s;

		auto r = regex(pattern,"gm");

		if (match(str,r)) return new Value(true);
		else return new Value(false);
	}

	static Value getMatches(Value[] v)
	{
		string str = v[0].content.s;
		string pattern = v[1].content.s;

		Value[] ret;

		auto r = regex(pattern,"gm");

		foreach(m; match(str,r)) 
		{
			string hit = m.hit;

			ret ~= new Value(hit);
		}

		return new Value(ret);
	}

	static Value preMatch(Value[] v)
	{
		string str = v[0].content.s;
		string pattern = v[1].content.s;

		auto r = regex(pattern,"gm");
		auto m = match(str,r);
		auto c = m.captures;

		string ret = to!string(c.pre);

		return new Value(ret);
	}

	static Value postMatch(Value[] v)
	{
		string str = v[0].content.s;
		string pattern = v[1].content.s;

		auto r = regex(pattern,"gm");
		auto m = match(str,r);
		auto c = m.captures;

		string ret = to!string(c.post);

		return new Value(ret);
	}

	static Value xreplaceFirst(Value[] v)
	{
		string str = v[0].content.s;
		string pattern = v[1].content.s;
		string repl = v[2].content.s;

		auto r = regex(pattern,"gm");
		string ret = std.regex.replaceFirst(str, r, repl);

		return new Value(ret);
	}

	static Value replaceAll(Value[] v)
	{
		string str = v[0].content.s;
		string pattern = v[1].content.s;
		string repl = v[2].content.s;

		auto r = regex(pattern,"gm");
		string ret = std.regex.replaceAll(str, r, repl);

		return new Value(ret);
	}

	static Value regexSplit(Value[] v)
	{
		string str = v[0].content.s;
		string pattern = v[1].content.s;

		auto r = regex(pattern,"gm");
		string[] parts = splitter(str, r).array();

		Value[] ret;

		foreach (string part; parts)
			ret ~= new Value(part);

		return new Value(ret);
	}
}
