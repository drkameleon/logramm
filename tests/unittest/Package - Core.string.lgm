import core.string;

out "hello".ucase();
out "HELLO".lcase();
out "    HELLO    ".trim();
out "    HELLO    ".ltrim();
out "    HELLO    ".rtrim();
out "HELLO".find("LL");
out "hello".chop();
out "hello".isNumeric();
out "12345".isNumeric();
out "hello".soundex();
out "hello".startsWith("he");
out "hello".startsWith("llo");
out "hello".endsWith("he");
out "hello".endsWith("llo");
out "hello\n".chomp();
out "hello".justify(10);
out "hello".justifyRight(10);
out "hello".justifyCenter(10);
out "hello".isEmpty();
out "".isEmpty();
out "hello".ucaseFirst();
out "HELLO".lcaseFirst();