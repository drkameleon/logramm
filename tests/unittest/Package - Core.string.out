HELLO
hello
HELLO
HELLO    
    HELLO
2
hell
false
true
H400
true
false
false
true
hello
hello     
     hello
  hello   
false
true
Hello
hELLO
